jansky/app.py

```diff
@@ -220,12 +211,58 @@ def __call__(self, key):
         return self.config(key)
 
 
+WorkItem = namedtuple('WorkItem', ['context', 'crash_id'])
+
+
+class Worklist:
+    """Generates crash ids to work on
+
+    If it's exhausted, by default it'll sleep for ``sleep_when_exhausted`` seconds and then try
+    again. If ``sleep_when_exhausted`` is <= 0, then it'll return.
+
+    """
+    def __init__(self, generator, sleep_when_exhausted=2):
+        self.generator = generator
+        self.sleep_when_exhausted = sleep_when_exhausted
+
+    def __iter__(self):
+        """Generates crash ids and completers"""
+        while True:
+            try:
+                workitem = self.generator.get_next()
+
+            except Exception:
+                # FIXME(willkg): the generator should handle all errors and throw a serviceerror for
+                # the errors it handled. anything else should be raised as a "real error" here.
+                pass
+
+            if workitem is not None:
+                yield workitem
+
+            elif self.sleep_when_exhausted > 0:
+                time.sleep(self.sleep_when_exhausted)
+
+            else:
+                return
+
+
```

Comment:
I'm just sketching things out at the moment. Trying to figure out a nice API. Figure I'll land this for now.